projects/plugins/jetpack/extensions/blocks/send-a-message/whatsapp-button/settings.js

```diff
@@ -0,0 +1,200 @@
+import {
+	BaseControl,
+	TextControl,
+	TextareaControl,
+	SelectControl,
+	Path,
+	ToggleControl,
+	Button,
+	PanelBody,
+	ToolbarGroup,
+	Dropdown,
+} from '@wordpress/components';
+import { PanelColorSettings } from '@wordpress/block-editor';
+import { DOWN } from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import { countryCodes } from '../shared/countrycodes.js';
+import renderMaterialIcon from '../../../shared/render-material-icon';
+import HelpMessage from '../../../shared/help-message';
+
+const WHATSAPP_GREEN = '#25D366';
+const WHATSAPP_DARK = '#465B64';
+const WHATSAPP_LIGHT = '#F4F4F4';
+
+export default function WhatsAppButtonSettings( { attributes, setAttributes, context } ) {
+	const { countryCode, phoneNumber, firstMessage, openInNewTab, backgroundColor } = attributes;
+
+	const [ isValidPhoneNumber, setIsValidPhoneNumber ] = useState( true );
+
+	const validatePhoneNumber = newPhoneNumber => {
+		// No alphabetical characters but allow dots, dashes, and brackets.
+		// These will be stripped for the WhatsApp API (only numbers), but retain
+		// them in the UI for a more readable number for the user.
+		const phoneNumberRegEx = RegExp( /^[+]?[\s./0-9]*[(]?[0-9]{1,4}[)]?[-\s./0-9]*$/, 'g' );
+
+		if ( undefined === newPhoneNumber || newPhoneNumber.length < 1 ) {
+			return false;
+		}
+
+		return phoneNumberRegEx.test( countryCode.replace( /\D/g, '' ) + newPhoneNumber );
+	};
+
+	const setBackgroundColor = color => {
+		setAttributes( { backgroundColor: color } );
+
+		if ( color === undefined || color === WHATSAPP_GREEN || color === WHATSAPP_DARK ) {
+			return setAttributes( { colorClass: 'dark' } );
+		}
+
+		setAttributes( { colorClass: 'light' } );
+	};
+
+	const renderSettingsToggle = ( isOpen, onToggle ) => {
+		const openOnArrowDown = event => {
+			if ( ! isOpen && event.keyCode === DOWN ) {
+				event.preventDefault();
+				event.stopPropagation();
+				onToggle();
+			}
+		};
+
+		return (
+			<Button
+				className="components-toolbar__control jetpack-contact-form__toggle"
+				label={ __( 'WhatsApp Button Settings', 'jetpack' ) }
+				onClick={ onToggle }
+				onKeyDown={ openOnArrowDown }
+				icon={ renderMaterialIcon(
+					<Path d="M10.82 12.49c.02-.16.04-.32.04-.49 0-.17-.02-.33-.04-.49l1.08-.82c.1-.07.12-.21.06-.32l-1.03-1.73c-.06-.11-.2-.15-.31-.11l-1.28.5c-.27-.2-.56-.36-.87-.49l-.2-1.33c0-.12-.11-.21-.24-.21H5.98c-.13 0-.24.09-.26.21l-.2 1.32c-.31.12-.6.3-.87.49l-1.28-.5c-.12-.05-.25 0-.31.11l-1.03 1.73c-.06.12-.03.25.07.33l1.08.82c-.02.16-.03.33-.03.49 0 .17.02.33.04.49l-1.09.83c-.1.07-.12.21-.06.32l1.03 1.73c.06.11.2.15.31.11l1.28-.5c.27.2.56.36.87.49l.2 1.32c.01.12.12.21.25.21h2.06c.13 0 .24-.09.25-.21l.2-1.32c.31-.12.6-.3.87-.49l1.28.5c.12.05.25 0 .31-.11l1.03-1.73c.06-.11.04-.24-.06-.32l-1.1-.83zM7 13.75c-.99 0-1.8-.78-1.8-1.75s.81-1.75 1.8-1.75 1.8.78 1.8 1.75S8 13.75 7 13.75zM18 1.01L8 1c-1.1 0-2 .9-2 2v3h2V5h10v14H8v-1H6v3c0 1.1.9 2 2 2h10c1.1 0 2-.9 2-2V3c0-1.1-.9-1.99-2-1.99z" />
```

Comment:
While we are refactoring, would this icon path be a candidate to move to its own file?

Reply:
I was trying to keep the refactoring down to just the parts required for better testing so the PR doesn't get too big, and cluttered with changes that don't affect testing ... but would be good to include in the follow up PR that looks at the dropdown sizing issue for sure.